dev/com.ibm.ws.jca_fat_regr/fat/src/suite/r80/base/jca16/ann/AnnUtils.java

```diff
@@ -0,0 +1,749 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2022 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package suite.r80.base.jca16.ann;
+
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.util.Scanner;
+
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.xpath.XPath;
+import javax.xml.xpath.XPathConstants;
+import javax.xml.xpath.XPathExpression;
+import javax.xml.xpath.XPathFactory;
+
+import org.w3c.dom.Document;
+import org.w3c.dom.NamedNodeMap;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.InputSource;
+
+import componenttest.topology.impl.LibertyServer;
+
+/**
+ *
+ */
+public class AnnUtils {
+    @SuppressWarnings("resource")
+    // the scanner is closed, the tooling thinks that it isn't because of the useDelimiter call...
+    public String getMetatype(LibertyServer server, String rar) throws Exception {
+
+        String rarName = "factoryPid=\"com.ibm.ws.jca.resourceAdapter.properties."
+                         + rar;
+        String traceFile = server.getLogsRoot() + "trace.log";
+        File file = new File(traceFile);
+        StringBuffer trace_log = new StringBuffer();
+        Scanner s = null;
+        try {
+            s = new Scanner(file).useDelimiter("\n");
+            boolean capture = false;
+            while (s.hasNext()) {
+                String line = s.next();
+                if (line.contains("<metatype"))
+                    capture = true;
+                if (capture) {
+                    trace_log.append(line);
+                }
+                if (line.contains(":MetaData>")) {
+
+                    String metatype = trace_log.toString().trim();
+                    if (metatype.contains(rarName + "\"")) {
+                        System.out.println("**************** METATYPE is: *****************" + metatype);
+                        return metatype;
+                    }
+                    trace_log = new StringBuffer();
+                    capture = false;
+                }
+            }
+        } finally {
+            if (s != null) {
+                s.close();
+                s = null;
+            }
+        }
+
+        return null;
```

Comment:
It looks like this method is reading every line of trace!  
This may be a really big performance hit!

Depending on how this method is used we could try one of the following options: 

1. If our tests are sequential. Meaning the first test does (x) and then gets metadata, and then the second test does (y) and then gets metadata then we don't need to search the entire trace, just from the end of the last test, to now.  For this, you could use `server.findStringsInLogsAndTraceUsingMark(regex)`. 
2. If our tests are non-sequential. Meaning the server starts and both (x) and (y) are just searching logs from the server startup (or something). Then we could use `findStringsInTrace(regex)` This will return a list containing all metatypes and we'd then have to parse out which metatype is for which test (which will be quicker then searching the logs every time.

Reply:
I took a look at those methods and did some experimenting and found and they won't work in this instance because they are looking for instances of the regex within single lines. The metatype definitions span multiple lines so what it does is find the line starting the metatype def ("<metatype") and then captures all the subsesquent lines until it finds a line with ":MetaData>" and all the lines between the start and end are appended into 1 multi-line metatype string. I looked to see if there was an existing method that did what we want using marks, and did not find one.

Reply:
Thanks for looking into this.  Let's consider improving the performance of this method as a stretch goal that we can do during the further development of the connectors-2.1 feature. Overall, since this test is just moving, we aren't negatively affecting performance as it is written currently.